### PR TITLE
Issue-26 Recycler View performance enhancement with DiffUtil Callback

### PR DESCRIPTION
**Pull request by Ivan2001otp**
## Fixes Issue
Improving the performance of recycler view by diffutil callback.I have also tried to explain with the video link given below.

## Demo video

[https://drive.google.com/file/d/1HRNRrXmmfjBhEm_EA0O58ElhKkCBRzEE/view?usp=sharing](url)

### CODE REVIEW

File: app/src/main/java/com/itssuryansh/taaveez/Notes.kt [hunk 1]
```diff
@@ -26,32 +26,46 @@ import com.itssuryansh.taaveez.databinding.ActivityNotesBinding
...
 import java.text.SimpleDateFormat
 import java.util.*
+import kotlin.collections.ArrayList
 
 
 class Notes : AppCompatActivity() {
 
     private var binding: ActivityNotesBinding? = null
     private var PoemDesUpdate: RichEditor ?=null
+    private var adapter_:itemAdapter ?= null//creating an instance of ItemAdapter
```
Comment (Vaibhav2002) [inline app/src/main/java/com/itssuryansh/taaveez/Notes.kt:42]: prefer using lateinit var

File: app/src/main/java/com/itssuryansh/taaveez/itemAdapter.kt [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
 import android.view.ViewGroup
+import androidx.recyclerview.widget.DiffUtil
 import androidx.recyclerview.widget.RecyclerView
+import com.itssuryansh.taaveez.Tools.RecyclerViewDiffUtil
 import com.itssuryansh.taaveez.databinding.ItemPoemBinding
 import kotlin.collections.ArrayList
 
 
-class itemAdapter(private val items: ArrayList<NotesEntity>,
+class itemAdapter(private var oldItems: ArrayList<NotesEntity>,
```
Comment (Vaibhav2002) [inline app/src/main/java/com/itssuryansh/taaveez/itemAdapter.kt:13]: Use `RecyclerView.ListAdapter`

File: app/src/main/java/com/itssuryansh/taaveez/Tools/RecyclerViewDiffUtil.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package com.itssuryansh.taaveez.Tools
+
+import androidx.recyclerview.widget.DiffUtil
+import com.itssuryansh.taaveez.NotesEntity
+
+class RecyclerViewDiffUtil(private val newList: ArrayList<NotesEntity>
```
Comment (Vaibhav2002) [inline app/src/main/java/com/itssuryansh/taaveez/Tools/RecyclerViewDiffUtil.kt:6]: Use `DiffUtil.ItemCallback`

File: app/src/main/java/com/itssuryansh/taaveez/Notes.kt [hunk 4]
```diff
@@ -196,12 +214,22 @@ class Notes : AppCompatActivity() {
...
                 },
             )
-
+        }
+        if (NotesList.isNotEmpty()) {
+            //this if-condition is invoked if the list is not empty.
+            /*
+            * Also , this if condition is also invoked for every new update in the notesList.
+            * */
             binding?.rvItemsPoem?.layoutManager = LinearLayoutManager(this)
```
Comment (Vaibhav2002) [inline app/src/main/java/com/itssuryansh/taaveez/Notes.kt:223]: dont repeat this step all the time, rather create a function to initialize RecyclerView
Inside it
- Initialize adapter, you wont need to pass List of Notes, when you use `RecyclerView.ListAdapter`
- set up recyclerView

File: app/src/main/java/com/itssuryansh/taaveez/itemAdapter.kt [hunk 5]
```diff
@@ -56,6 +59,17 @@ class itemAdapter(private val items: ArrayList<NotesEntity>,
     }
 
     override fun getItemCount(): Int {
-       return items.size
+       return oldItems.size
+    }
+
+    fun initiateDiffUtilCallback(newItems:ArrayList<NotesEntity>){
```
Comment (Vaibhav2002) [inline app/src/main/java/com/itssuryansh/taaveez/itemAdapter.kt:65]: Wont need this in `RecyclerView.ListAdapter`

File: app/src/main/java/com/itssuryansh/taaveez/NotesDao.kt [hunk 6]
```diff
@@ -21,4 +21,9 @@ interface NotesDao {
     @Query("select * from `Poem-table` where id=:id")
     fun fetchNotesById(id:Int):Flow<NotesEntity>
 
+
+    //used in diffutil callback implementation of recycler view
+    @Query("select * from `Poem-table`")
+    suspend fun getAllNotesSynchronously():List<NotesEntity>
```
Comment (Vaibhav2002) [inline app/src/main/java/com/itssuryansh/taaveez/NotesDao.kt:27]: Why do you need this?

**Review by Vaibhav2002 — CHANGES_REQUESTED** [review]
Please make the required changes

**Comment by Ivan2001otp** [comment]
Sure I will fix the code as per you mentioned .Thanks for your constructive
feedback.

**Pull request closed without merging**